providers/sru/bin/dkms_build_validation.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+#
+# Copyright 2017-2024 Canonical Ltd.
+# Written by:
+#   Taihsiang Ho (tai271828) <taihsiang.ho@canonical.com>
+#   Fernando Bravo <fernando.bravo.hernandez@canonical.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections import Counter
+import logging
+from packaging import version
+import re
+import subprocess
+import sys
+import textwrap
+from typing import Dict, List
+
+logger = logging.getLogger("dkms_build_validation")
+
+
+def run_command(command: str) -> str:
+    """Run a shell command and return its output"""
+    try:
+        result = subprocess.check_output(
+            command,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            universal_newlines=True,
+        )
+        return result.strip()
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(
+            "Command '{0}' failed with error:\n{1}".format(command, e.output)
+        )
+
+
+def parse_dkms_status(dkms_status: str, ubuntu_release: str) -> List[Dict]:
+    """Parse the output of 'dkms status' and return a list of dictionaries"""
+    kernel_info = []
+    for line in dkms_status.splitlines():
+        details, status = line.split(": ")
+        if version.parse(ubuntu_release) >= version.parse("22.04"):
+            kernel_ver = details.split(", ")[1]
+        else:
+            kernel_ver = details.split(", ")[2]
+        kernel_info.append({"version": kernel_ver, "status": status})
+    return kernel_info
+
+
+def parse_version(ver: str) -> version.Version:
+    """Parse the version string and return a version object"""
+    match = re.match(r"(\d+\.\d+\.\d+(-\d+)?)", ver)
+    if match:
+        parsed_version = version.parse(match.group(1))
+    else:
+        raise ValueError("Invalid version string: {0}".format(ver))
+    return parsed_version
+
+
+def has_dkms_build_errors(kernel_ver_current: str) -> bool:
+    log_path = "/var/log/apt/term.log"
+    err_msg = "Bad return status for module build on kernel: {}".format(
+        kernel_ver_current
+    )
+    with open(log_path, "r") as f:
+        for line in f.readlines():
+            if err_msg in line:
+                logger.error(
+                    "Found dkms build error messages in {}".format(log_path)
+                )
+                logger.error("\n=== build log ===")
+                result = run_command(
+                    "grep '{}' {} -C 5".format(err_msg, log_path)
```

Comment:
why do we need to run grep here?

Reply:
Tbh, I just left the grep there so I don't have to implement a functionality to get the context around each line. I will remove the grep.

Reply:
It should be possible to handle the context lines (the lines arround the error match). I saw that there is a grep-like implementation here that handles the context, but I don't think it is worth the effort to implement it here just for this small test. Maybe it is better to have the "/var/log/apt/term.log" as an attatchment.
If we think it is required, I could try to add it as a separate function.